x/auth/middleware/branch_store_test.go

```diff
@@ -0,0 +1,120 @@
+			priv1, _, addr1 := testdata.KeyTestPubAddr()
+			err := s.app.BankKeeper.MintCoins(ctx, minttypes.ModuleName, feeAmount)
+			s.Require().NoError(err)
+			err = s.app.BankKeeper.SendCoinsFromModuleToAccount(ctx, minttypes.ModuleName, addr1, feeAmount)
+			s.Require().NoError(err)
+			s.Require().Equal(feeCoin.Amount, s.app.BankKeeper.GetBalance(ctx, addr1, feeCoin.Denom).Amount)
+			seq, _ := s.app.AccountKeeper.GetSequence(ctx, addr1)
+			s.Require().Equal(uint64(0), seq)
+
+			// testMsgTxHandler is a test txHandler that handles one single TestMsg,
+			// consumes the given `tc.gasToConsume`, and sets the bank store "ok" key to "ok".
+			var testMsgTxHandler = customTxHandler{func(ctx context.Context, req tx.Request) (tx.Response, error) {
+				msg, ok := req.Tx.GetMsgs()[0].(*testdata.TestMsg)
+				if !ok {
+					return tx.Response{}, fmt.Errorf("expected %T, got %T", (*testdata.TestMsg)(nil), msg)
+				}
+
+				sdkCtx := sdk.UnwrapSDKContext(ctx)
+				sdkCtx.KVStore(s.app.GetKey("bank")).Set([]byte("ok"), []byte("ok"))
+				sdkCtx.GasMeter().ConsumeGas(tc.gasToConsume, "TestMsg")
+				if tc.panicTx {
+					panic("panic in tx execution")
+				}
+				return tx.Response{}, nil
+			}}
+
+			txHandler := middleware.ComposeMiddlewares(
+				testMsgTxHandler,
+				middleware.NewTxDecoderMiddleware(s.clientCtx.TxConfig.TxDecoder()),
+				middleware.GasTxMiddleware,
+				middleware.RecoveryTxMiddleware,
+				middleware.DeductFeeMiddleware(s.app.AccountKeeper, s.app.BankKeeper, s.app.FeeGrantKeeper),
+				middleware.IncrementSequenceMiddleware(s.app.AccountKeeper),
+				middleware.WithBranchedStore,
+				middleware.ConsumeBlockGasMiddleware,
+			)
+
+			// msg and signatures
+			msg := testdata.NewTestMsg(addr1)
+			var gasLimit uint64 = math.MaxUint64 // no limit on sdk.GasMeter
+			s.Require().NoError(txBuilder.SetMsgs(msg))
+			txBuilder.SetFeeAmount(feeAmount)
+			txBuilder.SetGasLimit(gasLimit)
+
+			privs, accNums, accSeqs := []cryptotypes.PrivKey{priv1}, []uint64{0}, []uint64{0}
+			testTx, _, err := s.createTestTx(txBuilder, privs, accNums, accSeqs, ctx.ChainID())
+			s.Require().NoError(err)
+
+			_, err = txHandler.DeliverTx(sdk.WrapSDKContext(ctx), tx.Request{Tx: testTx})
+
+			bankStore := ctx.KVStore(s.app.GetKey("bank"))
+			okValue := bankStore.Get([]byte("ok"))
+
+			if tc.expErr {
+				s.Require().Error(err)
+				if tc.panicTx {
+					s.Require().True(sdkerrors.IsOf(err, sdkerrors.ErrPanic))
+				} else {
+					s.Require().True(sdkerrors.IsOf(err, sdkerrors.ErrOutOfGas))
+				}
+				s.Require().Empty(okValue)
+			} else {
+				s.Require().NoError(err)
+				s.Require().Equal([]byte("ok"), okValue)
+			}
+			// block gas is always consumed
+			baseGas := uint64(24564) // baseGas is the gas consumed by middlewares
+			s.Require().Equal(tc.gasToConsume+baseGas, ctx.BlockGasMeter().GasConsumed())
+			// tx fee is always deducted
+			s.Require().Equal(int64(0), s.app.BankKeeper.GetBalance(ctx, addr1, feeCoin.Denom).Amount.Int64())
```

Comment:
I think, something is wrong here. We started with `150atom`, why it should always be zero?

We should test here that the proper amount of fee coins are deduced - this depends on the amount of gas used.

Reply:
Because the fee in tx is set to a constant for simplicity, `fee >= gasPrice * gasLimit` should be checked in checkTx, which is not tested here.

Reply:
But my question is about the tokens, not the gas. In line 47 we sent 150atoms to the user. Then after the transaction succeeds, some tokens should be debited to pay for the gas. So at the end we should still has some tokens.

Reply:
Because we specified the 150atom as the tax fee, it deduct exactly the amount of fee specified in tx.

Reply:
OK :+1: